File: Done/Binary_Search_307_Questions/36/Solution.java
class Solution {
    public int findRadius(int[] houses, int[] heaters) {
        Arrays.sort(houses);
        Arrays.sort(heaters);
        int n = houses.length;
        int m = heaters.length;
        int r = Math.max(heaters[0] - houses[0], houses[n-1] - heaters[m-1]);
        int j = 0;

        for (int i = 1; i < m; i++) {
            if (heaters[i-1]+r >= heaters[i]-r) {
                continue;
            }
            while (j < n && houses[j] <= heaters[i-1]+r) j++;

            while (j < n && houses[j] > heaters[i-1]+r && houses[j] < heaters[i]-r) {
                r += Math.min(houses[j] - (heaters[i-1]+r), (heaters[i]-r) - houses[j]);
                j++;
            }
        }

        return r;
    }
}

File: Done/Binary_Search_307_Questions/36/Solution.py
class Solution:
    def findRadius(self, houses: List[int], heaters: List[int]) -> int:
        houses.sort()
        heaters.sort()

        j = 0
        max_rad = 0
        for i in range(len(houses)):
            rad = abs(houses[i] - heaters[j])
           
            while j < len(heaters) - 1:
                next_rad = abs(houses[i] - heaters[j + 1])
                if next_rad <= rad:
                    j += 1
                    rad = next_rad
                else:
                    break
            
            max_rad = max(max_rad, rad)

        return max_rad